### ISSUE TITLE
Automation: Azure Pipelines has massive Promise leak, whereas local execution does not

### ISSUE THREAD

**Issue by headconnect** [pulumi/pulumi#6633]
When execution a pulumi automation script locally, either with `ts-node` or `tsc`, it executes and returns perfectly fine.
The stack is set up with remote secrets (azure keyvault) and remote state (azure storage account). 

Running the same automation with `tsc` (haven't reliably been able to get ts-node working at all yet) in Azure DevOps Pipelines results in:

```
Error: The Pulumi runtime detected that 904 promises were still active
at the time that the process exited. There are a few ways that this can occur:
  * Not using `await` or `.then` on a Promise returned from a Pulumi API
  * Introducing a cyclic dependency between two Pulumi Resources
  * A bug in the Pulumi Runtime
```

## Expected behavior

The run of `automation.ts` should execute the same on the pipeline agent as locally.

## Current behavior

Local environment: 
```
win10
λ pulumi version
2.23.2
λ node --version
v15.11.0
λ tsc --version
Version 4.2.3
```
Remote environment:
```
Ubuntu 20.04 Pipeline Agent
pulumi: v2.23.1
node: 14.16.0
tsc: Version 4.2.3 
```
Both environments execute `stack.info()` without a problem, returning the same information.

Running `stack.preview()` is where the difference between environments is shown:
Execution locally:
```
...
  resourceChanges: { delete: 11, same: 12 }
}
previewing stack
<.... (stuff here) ...>
Resources:
    + 1 to create
    - 1 to delete
    2 changes. 2 unchanged
``` 

Whereas the pipeline version of the exact same thing returns:

```
...
  resourceChanges: { delete: 11, same: 12 }
}
previewing stack
Error: The Pulumi runtime detected that 904 promises were still active
at the time that the process exited. There are a few ways that this can occur:
  * Not using `await` or `.then` on a Promise returned from a Pulumi API
  * Introducing a cyclic dependency between two Pulumi Resources
  * A bug in the Pulumi Runtime
```

When running with PULUMI_DEBUG_PROMISE_LEAKS, literally hundreds of these (slightly anonymized as taken from original run):
```
CONTEXT(664): transferIsStable(resource:zone8-node1234[azure-native:network:Subnet], networkSecurityGroup, [object Object])
STACK_TRACE:
Error
    at Object.debuggablePromise (/agent/_work/1/s/pulumi/<projectpath>/node_modules/@pulumi/pulumi/runtime/debuggable.js:69:75)
    at Object.transferProperties (/agent/_work/1/s/pulumi/<projectpath>/node_modules/@pulumi/pulumi/runtime/rpc.js:68:189)
    at /agent/_work/1/s/pulumi/<projectpath>/node_modules/@pulumi/pulumi/runtime/resource.js:361:33
    at Generator.next (<anonymous>)
    at /agent/_work/1/s/pulumi/<projectpath>/node_modules/@pulumi/pulumi/runtime/resource.js:21:71
    at new Promise (<anonymous>)
    at __awaiter (/agent/_work/1/s/pulumi/<projectpath>/node_modules/@pulumi/pulumi/runtime/resource.js:17:12)
    at prepareResource (/agent/_work/1/s/pulumi/<projectpath>/node_modules/@pulumi/pulumi/runtime/resource.js:331:12)
    at Object.registerResource (/agent/_work/1/s/pulumi/<projectpath>/node_modules/@pulumi/pulumi/runtime/resource.js:213:24)
    at new Resource (/agent/_work/1/s/pulumi/<projectpath>/node_modules/@pulumi/pulumi/resource.js:215:24)
```

Where CONTEXT is repeated for the same thing multiple times, with minor variations, starting from resolveURN and resolveURNIsKnown and resolveID through a whole bunch of transferDeps, transferProperty, transferIsStable, transferSecret until just the resource itself is in context.

For that resource (the subnet) there are 107 contexts enumerated like above.

## Steps to reproduce

See github repo - though for it to work you will likely need to use remote state for it to work at all.
https://github.com/headconnect/pulumi-bug-reproducer/tree/issue/automation-6633-promise-leak

## Context (Environment)

Prevents execution of normal Pulumi stacks through automation in pipelines, which would allow me to do quite a lot of really cool stuff.

**Comment by barclayadam** [comment]
This is still an issue for me too sometimes, even though #5853 had a fix applied for it. In particular, this occurs for me when an error occurs in the `up` command.

@headconnect I applied a workaround that was mentioned in #6153 with regards to returning the id of the AKS provider I register, that may help you to get working stacks running at least

**Comment by headconnect** [comment]
I'm not certain I'll be able to deal with that for all secrets - but the output itself is so incredibly massive, with promise leaks on essentially every single resource in my entire code, that I'm not sure this is a viable fix. Again - it's 100% reproducible on azure pipelines and 100% working when running locally. I'm almost thinking there could be some environmental differences (i.e. local variables or whatnot) causing this. 

I could try to run it on a windows agent just to see if that changes anything.

**Comment by barclayadam** [comment]
I had pretty much exactly the same for a period of time, I could never reproduce the failures locally... Until one day the promise leak did start happening locally. I never tracked down what made the difference, but my guess is this is a race condition that presented itself more readily in certain environments / for certain stacks.

The workaround is a (potentially) one-liner. The change is to output a provider ID with your outputs. It's not per-secret or resource, but a single value being output. In my case I grabbed ID from a K8S provider (i.e. as an output of the stack I have `$k8sProviderId` set to `new k8s.Provider(...).id`

**Comment by headconnect** [comment]
Tried that now - didn't work - same amount of leakage :/ 

Added aks.id to the output - tested locally - verified that it output correctly - ran exact same thing in azure pipeline - promise leaks to the moon.

Almost tempted to spin up a custom azure pipeline agent with 3.0.0-beta on it to see if something magically changes..

**Comment by hugbubby** [comment]
This behavior is rearing its head for me too. In my case I'm running the pulumi automation api on a kubernetes cluster. When I use telepresence to run the automation api script locally, it succeeds. When the same script runs inside a pod on the cluster, it fails, and I get an enormous amount of unhandled promise rejections.

This is the latest version of pulumi. It sounds like a race condition.

**Comment by EvanBoyle** [comment]
> this occurs for me when an error occurs in the up command
Opened an issue to track fixing this. We should skip the leak checker if you had error in your program: https://github.com/pulumi/pulumi/issues/6756

**Comment by EvanBoyle** [comment]
@hugbubby can you provide additional details about the script your running and the failures that you see? How are you running it locally, and how are you running it in your pod? How many and what kind of resources are you creating? Do you have debug logs that you can share or output from the failure?

**Comment by EvanBoyle** [comment]
Fix is out for running the promise leak checker on error: https://github.com/pulumi/pulumi/pull/6758

Curious to understand if there are any common factors here resource-wise. Are any of you using our helm support? @headconnect @barclayadam

**Comment by headconnect** [comment]
Looking forward to testing this out in the coming days :)

In the original issue I was not using any helm support, but I do have a separate project for k8s resources designed in a similar way in which I am using helm - and there have been some promise issues there as well (though not at the point where I've tried pipeline automation for that project yet). 

In other news, through discussion with Piers - I've come to realize that I have some good potential for some simple refactoring of my code which might address some of the promise issues. Essentially my code for the pulumi program has been similar to this:

```ts
export const pulumiProgram = async () => {
  let fullStack = {}

   let fullStack = await createFirstResource(fullStack)

   let fullStack = await createSecondResource(fullStack)

   ...etc...

  return fullStack
}
```

where `createNthResource()` is like this: 
```ts
export const creeateFirstResource = async (firstResource) => {
   // create 2-10 resources with ~400 loc of typescript and various stuff
  // also updating firstResrouce (fullStack) with resource.id and resource.name and possibly some config
  return firstResource
}
```

This is obviously a good pattern for ComponentResources which I was maliciously unaware of.. until now. So hoping to do some good refactoring the coming days - but I'll keep the old way to test.

**Comment by EvanBoyle** [comment]
The fix for skipping the leak checker on error has been released in 2.25.

**Comment by EvanBoyle** [comment]
fixed via https://github.com/pulumi/pulumi/issues/6756
